clang/lib/Sema/SemaDecl.cpp

```diff
@@ -10897,6 +10902,53 @@ Attr *Sema::getImplicitCodeSegOrSectionAttrForFunction(const FunctionDecl *FD,
   return nullptr;
 }
 
+// Should only be called when getFunctionEffects() returns a non-empty set.
+// Decl should be a FunctionDecl or BlockDecl.
+void Sema::maybeAddDeclWithEffects(const Decl *D,
+                                   const FunctionEffectsRef &FX) {
+  if (!D->hasBody()) {
+    if (const auto *FD = D->getAsFunction(); FD && !FD->willHaveBody())
+      return;
+  }
+
+  if (Diags.getIgnoreAllWarnings() ||
+      (Diags.getSuppressSystemWarnings() &&
+       SourceMgr.isInSystemHeader(D->getLocation())))
+    return;
+
+  if (hasUncompilableErrorOccurred())
+    return;
```

Comment:
This looks familiar. I think there is similar (or identical) code somewhere in `AnalysisBaseWarnings.cpp`; we should probably factor that out into a separate helper function.

Reply:
Something like this as a new method of `Sema`:

`bool shouldSkipComplexAnalysis(SourceLocation SL);`

?

Reply:
Yeah, something like that; maybe calling it `shouldSkipAnalysisBasedWarnings()` or sth like that would be better because that’s mainly what it’d be used for.

Reply:
Well, I extracted that function, searched for users of `hasUncompilableErrorOccurred` and discovered that they are all subtly different. The closest candidate in AnalysisBasedWarnings has this extra twist on the pattern:

```
  if (S.hasUncompilableErrorOccurred()) {
    // Flush out any possibly unreachable diagnostics.
    flushDiagnostics(S, fscope);
    return;
  }
```

Reply:
Ah yes, of course they are... Yeah, my bad, in that case, this can probably just stay as-is.